### PR TITLE
docs(tfhe): add fhe_strings example

### PR DESCRIPTION
**Pull request by JoseSK999**
This PR adds the FHE str API example.

### CODE REVIEW

File: tfhe/examples/fhe_strings/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# fhe_strings
+
+This repo contains the implementation of a str API in FHE, featuring 30 methods. This API allows the user to:
+* Encrypt the `str` with or without padding nulls (i.e. encrypted `0u8`s at the end of the string), which serve to obfuscate the length but are ignored by algorithms
```
Comment (IceTDrinker) [inline tfhe/examples/fhe_strings/README.md:4]: nitpick: Perhaps to update, as it's not a repo anymore, rather an example

File: tfhe/examples/fhe_strings/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+* Encrypt the number of repetitions `n`, allowing to provide a clear `max` to restrict the range of the encrypted `n`
+* Provide a cleartext pattern when algorithms can run faster. Otherwise, it's possible to trivially encrypt the pattern with `FheString::trivial`
+
+Encrypted strings contain a flag indicating whether they have padding nulls or not. Algorithms are optimized to differentiate between the two kind of strings. For instance, in some cases we can skip entirely the FHE computations if we know the true lengths of the string or pattern.
+
+Just like the clear str API, any encrypted string returned by a function can be used as input to other functions. For instance when `trim_start` is executed, or a `Split` iterator instance is advanced with `next`, the result will only have nulls at the end. The decryption function `decrypt_ascii` will panic if it encounters with malformed encrypted strings, including padding inconsistencies.
+
+We have handled corner cases like empty strings and empty patterns (with and without padding), the number of repetitions `n` (clear and encrypted) being zero, etc. A complete list of tests can be found at `src/assert_functions/test_vectors.rs`.
+
```
Comment (IceTDrinker) [inline tfhe/examples/fhe_strings/README.md:14]: maybe a small example demonstrating using some basic usage of the lib with maybe two chained methods could be nice and compare the result to the rust clear result

File: tfhe/examples/fhe_strings/assert_functions/test_vectors.rs [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+    ("a", 1),
+    ("o", 2),
+    (" ", 2),
+    ("I F*", 3),
+    ("<3", 3),
+    ("foo", 0),
+    ("foofoo", 0),
+    ("foofoo", 1),
+    (" zelda", 12),
+    ("I F*** <3 zelda", 0),
```
Comment (IceTDrinker) [inline tfhe/examples/fhe_strings/assert_functions/test_vectors.rs:18]: hmm, maybe something a bit less edgy here please ;)

File: tfhe/examples/fhe_strings/assert_functions/test_vectors.rs [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+const TEST_CASES_MATCH: [(&str, u32); 15] = [
+    ("", 0),
+    ("", 1),
+    ("", 2),
+    ("", 3),
+    ("a", 0),
+    ("a", 1),
+    ("o", 2),
+    (" ", 2),
+    ("I F*", 3),
```
Comment (IceTDrinker) [inline tfhe/examples/fhe_strings/assert_functions/test_vectors.rs:12]: same here something a bit less edgy please :)

File: tfhe/examples/fhe_strings/README.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
+# fhe_strings
```
Comment (IceTDrinker) [inline tfhe/examples/fhe_strings/README.md:1]: title:

FHE Strings

File: tfhe/examples/fhe_strings/README.md [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
```
Comment (IceTDrinker) [inline tfhe/examples/fhe_strings/README.md:2]: I think you use an interesting method by "converting" a string to a uint in some algorithms (if I understand correctly) it could be worth explaining the approach and the intuition behind that transformation in this readme, a small paragraph could be nice

**Review by IceTDrinker — COMMENTED** [review]
Hello @JoseSK999 the PR is fairly large, so we will be reviewing in chunks, I did some of the high level nitpicks, mainly on the readme, I will try to take a look at the docstrings and some of my colleagues will check the algorithms and test cases in more depth

Generally the code is clean and is one of the reasons you are a winner, still we'll likely have some comments regarding some stuff, we have been pretty busy, so sorry for the delay !

**Review by IceTDrinker — COMMENTED** [review]
Some additional minor stuff for docstrings, but looks good overall for documentation !

File: tfhe/examples/fhe_strings/server_key/comp.rs [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+    ///
+    /// Returns `false` if they are not equal.
+    ///
+    /// The pattern for comparison (`rhs`) can be specified as either `GenericPattern::Clear` for a
+    /// clear string or `GenericPattern::Enc` for an encrypted string.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// let (ck, sk) = gen_keys();
```
[the remark this answers is not in the document]
Comment (JoseSK999) [reply tfhe/examples/fhe_strings/server_key/comp.rs:56]: I think docstrings are not compiled as tests because this is actually a binary, and doctests seem to only work in libraries

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/examples/fhe_strings/server_key/comp.rs:56]: yep, I would still add the imports as this example will likely get in the lib down the line in form or the other :)

File: tfhe/examples/fhe_strings/server_key/pattern/contains.rs [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+
+                    self.asciis_eq(a.into_iter(), b.into_iter())
+                }
+            })
+            .collect();
+
+        for match_case in matched {
+            // One of the possible values of pat must match the str
+            self.key.boolean_bitor_assign(&mut result, &match_case);
+        }
```
Comment (tmontaigu) [inline tfhe/examples/fhe_strings/server_key/pattern/contains.rs:40]: Actually here, you should be able to transform all these matched booleanblocks into a RadixCiphertext, and use scalar_ne_parallelized(&radix, 0) on it, it should be faster

File: tfhe/examples/fhe_strings/server_key/pattern/contains.rs [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+                let a: Vec<&FheAsciiChar> = str_chars.collect();
+
+                self.clear_asciis_eq(a.into_iter(), pat)
+            })
+            .collect();
+
+        for match_case in matched {
+            // One of the possible values of pat must match the str
+            self.key.boolean_bitor_assign(&mut result, &match_case);
+        }
```
Comment (tmontaigu) [inline tfhe/examples/fhe_strings/server_key/pattern/contains.rs:65]: Same as in `compare_shifted`

File: tfhe/examples/fhe_strings/server_key/pattern/mod.rs [hunk 10]
```diff
@@ -0,0 +1,209 @@
...
+// Useful for handling cases in which we know if there is or there isn't a match just by looking at
+// the lengths
+enum IsMatch {
+    Clear(bool),
+    Cipher(BooleanBlock),
+    None,
+}
+
+impl ServerKey {
+    fn length_checks(&self, str: &FheString, pat: &FheString) -> IsMatch {
```
Comment (tmontaigu) [inline tfhe/examples/fhe_strings/server_key/pattern/mod.rs:21]: This could use a small descriptions

File: tfhe/examples/fhe_strings/ciphertext.rs [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+            .collect();
+
+        let mut uint = RadixCiphertext::from_blocks(blocks);
+
+        if uint.blocks().is_empty() {
+            sk.key()
+                .extend_radix_with_trivial_zero_blocks_lsb_assign(&mut uint, 4);
+        }
+
+        uint
```
[the remark this answers is not in the document]
Comment (JoseSK999) [reply tfhe/examples/fhe_strings/ciphertext.rs:166]: Yess that makes sense

File: tfhe/examples/fhe_strings/server_key/pattern/contains.rs (hunk 8 above)
Comment (JoseSK999) [reply tfhe/examples/fhe_strings/server_key/pattern/contains.rs:40]: Thanks! Very useful

**Comment by JoseSK999** [comment]
Everything has been addressed. The docstring should now work in case they were compiled, some comments, docstrings and README updated, and the improvement in `len`, `compare_shifted`, `compare_shifted_clear` and `to_uint`.

Thanks!

**Comment by IceTDrinker** [comment]
@slab-ci cpu_fast_test
@slab-ci gpu_test

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test
@slab-ci cpu_unsigned_integer_test
@slab-ci cpu_signed_integer_test
@slab-ci cpu_wasm_test
@slab-ci csprng_randomness_testing

**Review by IceTDrinker — APPROVED** [review]

**Review by IceTDrinker — CHANGES_REQUESTED** [review]
Hmm actually @JoseSK999 it looks like there are test failures with the latest version of the code

e.g.

test assert_functions::test_vectors::test_split_ascii_whitespace ... FAILED
test assert_functions::test_vectors::test_trim ... FAILED
test assert_functions::test_vectors::test_trim_start ... FAILED

could you check on your end ?

**Comment by JoseSK999** [comment]
So the problem is that in the new version of `len` we assume a padded string to have at least one char (a null), which is correct.
```rust
let len = self
    .key
    .sum_ciphertexts_parallelized(non_zero_chars.iter())
    .expect("There's at least one padding character");
```

However in `trim_start` we set the trimmed result to padded to check the new length (in FHE to discard potential nulls). The diff between new and old lengths is the amount to shift the trimmed string left, to make it null-terminated.

So it panics because in `trim_start` we set an empty string as padded. Actually the `trim` functions and `split_ascii_whitespace` should return early in the empty str case. It's better for efficiency and also fixes this.

**Comment by JoseSK999** [comment]
Fixed

**Review by IceTDrinker — APPROVED** [review]
Thanks for the fix

**Comment by JoseSK999** [comment]
Clippy is happy now ;)
